bpf/protocol_http2.h

```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
     bpf_map_delete_elem(&ongoing_http2_grpc, stream);
 }
 
-static __always_inline void process_http2_grpc_frames(pid_connection_info_t *pid_conn,
-                                                      void *u_buf,
-                                                      int bytes_len,
-                                                      u8 direction,
-                                                      u8 ssl,
-                                                      u16 orig_dport) {
-    int pos = 0;
-    u8 found_start_frame = 0;
-    u8 found_end_frame = 0;
-
-    http2_grpc_request_t *prev_info = 0;
-    u32 saved_stream_id = 0;
-    int saved_buf_pos = 0;
-    u8 found_data_frame = 0;
-    http2_conn_stream_t stream = {0};
-    stream.pid_conn = *pid_conn;
-
-    unsigned char frame_buf[FRAME_HEADER_LEN];
-    frame_header_t frame = {0};
-
-    for (int i = 0; i < 8; i++) {
-        if (pos >= bytes_len) {
+static __always_inline frame_header_t next_frame(const grpc_frames_ctx_t *g_ctx) {
```

Comment:
The compiler will always optimise the return value here, even with -O0 - both gcc and clang will do it. It may not optimise with `-O0`  case for C (it will for C++ >=17 because of [mandatory copy elision)](https://en.cppreference.com/w/cpp/language/copy_elision) for the general case, but it has to and will for our particular case because of `__always_inline` - as a matter of fact, it won't even compile otherwise, as bpf does not support returning aggregates. Check out [this example](https://godbolt.org/z/4nYe8WY4r). So that guarantees it for us.

Passing things as out parameters should be avoided when possible because it prevents [const-correctness](https://isocpp.org/wiki/faq/const-correctness).

I would still like to keep it as is for clarity and the reasons above (const is something we should strive to use more, as it will shift a few bugs to compile time and increase clarity), even if it costs a few instructions - which actually will also not be the case. For no optimisation cases, the compiler will allocate the return value on the stack and then read from it - for optimised cases, the compiler will read directly from the return register if we return a `u8`, but will have to read from a stack offset in the present case, which is a negligible penalty, if anything (it's probably hot on cache anyway), which is offset in case we need to pass a in-out pointer anyway.

Reply:
_Update:_ I ended up removing the return value (and the extra enum) altogether as they were causing the loop to bail earlier and causing a bug. The correct behaviour, consistent with the pre-refactor code, is to continue looping into the next frame.

Reply:
Ok, great. I really think we should avoid relying on clang will optimize this away in a form which we can easily write.